src/lib.rs

```diff
@@ -42,10 +42,14 @@ extern crate libc;
 
 mod errors;
 
-use libc::{pid_t, c_char, c_void, getpid, size_t, c_int};
+use libc::{pid_t, c_char, c_void, getpid, size_t, c_int, geteuid};
 use std::ffi::CString;
 use std::path::PathBuf;
 use errors::TraceMeError;
+use std::fs::File;
+use std::io::Read;
+
+const PERF_PERMS_PATH: &'static str = "/proc/sys/kernel/perf_event_paranoid";
```

Comment:
I think this should be guarded with a `#[cfg(Linux)]` or whatever it is. At some point we'll want this run on non-Linux, so having a compile-time error will make it easier to port.

Reply:
Shall we just say this doesn't work on !Linux in the README? The whole thing relies on perf. Or put on test at an entry point or something?

Reply:
Let's make build.rs crash on !Linux.

Reply:
From experience of porting, it's *much* better if you make each individual point of likely divergence fail. Just saying "it doesn't work on !Linux" globally doesn't help a porter much.

Reply:
OK, but they will be everywhere.

Reply:
In fact, I'd like to do that as a separate PR, if that's OK?

Reply:
Well if they end up in too many places, maybe we need to rethink the code layout a bit?

Reply:
I've no problem with that being a separate PR.

Reply:
I think we should ease off the software engineer a little for now, since we have no idea if our ideas will work yet. I'm very keen to move on and it doesn't need to be rock solid just yet. We can introduce platform abstractions later.

Reply:
Point taken. But keep it in mind: it's really painful to introduce these abstractions to even a moderately sized project. So it might be better to bite the bullet at some point not ridiculously far into the future.

Reply:
#17.

You might be right actually. So I think we would want switchable backends, maybe a Travis backend too, which would somewhat address somewhat #10. Hrm. The perf backend basically consists of all of the current C code, so I'd even be tempted to rename traceme.c to `backend_perf.c` or something.

Let me integrate this into Rustc, then we can come back to this.